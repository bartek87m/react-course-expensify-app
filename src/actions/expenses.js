import uuid from 'uuid';
import database from '../firebase/firebase';
//Akcje wysyłamy do store poprzez store.dispatch

//AddExpense
export const addExpense = (expense) =>({
        type: 'ADD EXPENSE',
        expense
});

export const startAddExpense = (expenseData = {}) => {
     //zwracamy funkcję zamiast obiektu, działa tylko dzięki użyciu middleware redux thunk
    return (dispatch) => { //daje nam dostęp do dispatch
        const { //alternatywny sposób na zapisanie domyślnych wartości
            description = '', 
            note = '', 
            amount = 0, 
            createdAd = 0
        } = expenseData;
        
        const expense = {description, note, amount, createdAd}; //obiekt który przekazujemy 
        //zapytania do database są asynchroniczn i działają jako promisce
        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({//po dodaniu danych do firebase zapisujemy je do redux
                id: ref.key,
                ...expense
            }));
        });
    }
};

//Remove Expense
export const removeExpense = ({id} = {}) => ({
    type: 'REMOVE EXPENSE',
    id
});

//Edit Expense
export const editExpense = (id, updates) => ({
    type: 'EDIT EXPENSE',
    id,
    updates
})

