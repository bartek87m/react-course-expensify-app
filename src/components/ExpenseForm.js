import React from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';


// const date = new Date();
const now = moment();
console.log(now.format("MMM Do, YYYY"));

export default class ExpenseForm extends React.Component{
   constructor(props){//pobranie propsów
    super(props); //przekazujemy je do super
    console.log(props);
    this.state = {
        description: props.expense ? props.expense.description: '',
        note: props.expense ? props.expense.note: '',
        amount: props.expense ? (props.expense.amount / 100).toString(): '',
        createdAd: props.expense ? moment(props.expense.createdAd): moment(),
        calendarFocused: false,
        error: ''
    }

   }
 
  
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({description}))
    }

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({note}))
    }

    onAmountChange = (e) => {
        const amount = e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){//wyrażenie rególarne
             this.setState(() => ({amount}));
        }
    }

    onDataChange = (createdAd) => {
        if(createdAd) {
            this.setState(() => ({createdAd}))
        }
        
    }

    onFocusChange = ({focused}) => {
        this.setState(() => ({calendarFocused: focused}))
    }

    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({error: 'Please enter valid data'}));
        }else{
            this.setState(() => ({error: ''}));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAd: this.state.createdAd.valueOf(),
                note: this.state.note


            });
        }
    }

    render(){
        return(
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus //przy wejściu na stronę automatycznie ustawia kursor na ten input  
                        value={this.state.description}
                        onChange = {this.onDescriptionChange}
                        />
                    <input
                        type="number"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date = {this.state.createdAd}
                        onDateChange =  {this.onDataChange}
                        focused = {this.state.calendarFocused}
                        onFocusChange = {this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />

                    <textarea
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                        />
                    <button>Add Expense</button>
                    
                    </form>
            </div>
        )
    }
}