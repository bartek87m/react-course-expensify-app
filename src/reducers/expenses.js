
//Expenses Reducer

const expensesReducerDefault = [];

export default (state = expensesReducerDefault, action) => {
    switch (action.type) {
        case 'ADD EXPENSE':
            // return state.concat(action.expense) //concat bierze tablice dadaje element i zwraca nowÄ…
            return[
                ...state,action.expense
            ]
        case 'REMOVE EXPENSE':
            console.log("Action ID:" + action.id);
            return state.filter(({id}) => id !== action.id);
        
        case 'EDIT EXPENSE':
            return state.map((expense)=>{
                if (expense.id === action.id){
                    return{
                        ...expense,
                        ...action.updates
                    }
                }else{
                    return expense;
                }
            });

        case 'SET EXPENSES':
            return action.expenses;   
                  
        default:
            return state;
    }
}

