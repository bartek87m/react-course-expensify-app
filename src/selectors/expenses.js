
//Get visible expenses
import moment from 'moment';

export default (expenses, {text, sortBy, startDate, endDate}) => {

    return expenses.filter(expense => {
        const createdAdMoment = moment(expense.createdAd);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAdMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAdMoment, 'day') : true;


        const textMatch = (expense.description).toLowerCase().includes(text.toLowerCase());

        // console.log(textMatch);

        return startDateMatch && endDateMatch && textMatch;
            }).sort((a ,b) => {
        if(sortBy === 'date'){
            return a.createdAd < b.createdAd ? 1 : -1
        }
        else if(sortBy === 'amount'){
            return a.amount < b.amount ? 1: -1
        }
    
    });
}