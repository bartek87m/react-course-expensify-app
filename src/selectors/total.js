
export default (expenses) => {
    
    console.log("from total",expenses.length)
    if(expenses.length === 0){
        return 0;
    } else {
        return (expenses)
            .map(expense =>  expense.amount)
            .reduce((accumulator, currentValue) => accumulator + currentValue);
    }  


}