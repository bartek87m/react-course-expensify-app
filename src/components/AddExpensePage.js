import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from "./ExpenseForm";
import {startAddExpense} from '../actions/expenses';

export class AddExpensePage extends React.Component{
    onSubmit = (expense) => {
        // props.dispatch(addExpense(expense));
        console.log(expense)
        this.props.startAddExpense(expense)
        this.props.history.push('/');//tak jakby ktoś kliknoł link
    } 
    render() {
        return(
            <div>
            <h1>Add Expense</h1>
            <ExpenseForm
                onSubmit={this.onSubmit}
            />
        </div> 
        )
    }
}

const mapDispatchToProps = (dispatch) => ({ //pozwala zwrócić dispathc poza komponentem
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
    
})

export default connect(undefined, mapDispatchToProps)(AddExpensePage);