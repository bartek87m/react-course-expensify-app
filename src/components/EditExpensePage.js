import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm'
import {editExpense, removeExpense} from '../actions/expenses';

export class EditExpensePage extends React.Component{

    onSubmit = (expense) => {//to się dzieje kiedy ExpenseForm jest submitowana:)
        // props.dispatch(editExpense(props.expense.id, expense))
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push('/')//przekierowanie
    }

    onClick = () => {
        // props.dispatch(removeExpense({ id: props.expense.id}));
        // console.log("ID: " + this.props.expense.id);
        this.props.removeExpense(this.props.expense.id);
        this.props.history.push('/')//przekierowanie
    }

    render() {
        return(

            <div>
                <ExpenseForm
                    expense = {this.props.expense}//przekazanie danych uzyskanych z redux
                    onSubmit= {this.onSubmit}
                />
                <button onClick={this.onClick}>Remove</button>
            </div>
        )
        
    }
}

const mapStateToProps = (state, props) => ({
    
        expense:state.expenses.find((expense) => expense.id === props.match.params.id)
             //propsy z higher component
        
    
});

const mapDispatchToProps = (dispatch, props) => ({
    
        editExpense: (id, expense) => dispatch(editExpense(id, expense)),
        removeExpense: (data) => dispatch(removeExpense({id: data}))
        
}); 

//dzięki connect poprzez props mamy dostęp do dispatch
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage); //higher component