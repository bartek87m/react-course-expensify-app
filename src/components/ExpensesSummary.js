import React from 'react'
import { connect } from 'react-redux';
import total from "../selectors/total";
import selecrtedExpenses from '../selectors/expenses'
import numeral from 'numeral';
import { Link } from 'react-router-dom';

export const ExpensesSummary = (props) => {
       
   
    console.log("form Summary", props.visibleExpenses2);
        
        let expenseCount = props.visibleExpenses.length;
        let expensesTotal = numeral(props.expensesTotal / 100).format("$0,0.00");

        return (
            <div className="page-header">
                <div className="content-container">
                    <h1 className="page-header__title">Viewing <span>{expenseCount} </span>expenses totalling <span>{expensesTotal}</span></h1>
                    <div className="page-header__actions">
                        <Link className="button" to="/create">Add Expense</Link>
                    </div>
                </div>
            </div>
            )
    }
    

const mapStateProps = (state) => {
    const visibleExpenses = selecrtedExpenses(state.expenses, state.filters)
    return {
        visibleExpenses: selecrtedExpenses(state.expenses, state.filters),
        expensesTotal: total(visibleExpenses)
    };
}


export default connect(mapStateProps)(ExpensesSummary); //higher component