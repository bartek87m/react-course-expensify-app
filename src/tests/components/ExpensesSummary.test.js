import React from 'react';
import {shallow} from 'enzyme';
import {ExpensesSummary} from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses'
import total from '../../selectors/total'

test('should render ExpenseSummary with multiple expenses', () => {
    const wrapper = shallow(<ExpensesSummary visibleExpenses={expenses.length} expensesTotal = {total(expenses)}/>)
    expect(wrapper).toMatchSnapshot();
})

test('should render ExpenseSummary with one expense', () => {
    const wrapper = shallow(<ExpensesSummary visibleExpenses={expenses[1]} expensesTotal = {total([expenses[1]])}/>)
    expect(wrapper).toMatchSnapshot();
})
