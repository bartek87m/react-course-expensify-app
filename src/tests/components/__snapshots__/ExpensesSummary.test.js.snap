// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseSummary with multiple expenses 1`] = `
<div>
  <p>
    Viewing 
     expenses totalling 
    $1,141.95
  </p>
</div>
`;

exports[`should render ExpenseSummary with one expense 1`] = `
<div>
  <p>
    Viewing 
     expenses totalling 
    $1,095.00
  </p>
</div>
`;
